File: algorithm-design/java/src/test/java/leetcode/LC242ValidAnagram.java
package leetcode;

import org.junit.Test;

import java.util.Arrays;

import static org.junit.Assert.assertEquals;

/**
 * https://leetcode.com/problems/valid-anagram/
 */
public class LC242ValidAnagram {
    public class Solution {
        public boolean isAnagram(String s, String t) {
            char[] sChars = s.toCharArray();
            char[] tChars = t.toCharArray();
            if (sChars.length != tChars.length) {
                return false;
            }
            Arrays.sort(sChars);
            Arrays.sort(tChars);
            for (int i = 0; i < sChars.length; i++) {
                if (sChars[i] != tChars[i]) {
                    return false;
                }
            }
            return true;
        }
    }

    @Test
    public void test_a_b() throws Exception {
        assertEquals(false, new Solution().isAnagram("a", "b"));
    }

    @Test
    public void test_anagram_nagaram() throws Exception {
        assertEquals(true, new Solution().isAnagram("anagram", "nagaram"));
    }

    @Test
    public void test_rat_car() throws Exception {
        assertEquals(false, new Solution().isAnagram("rat", "car"));
    }
}

File: algorithms/code/leetcode/lc242_valid_anagram/lc242_valid_anagram.rs
// https://leetcode.com/problems/valid-anagram/
// #easy

pub struct Solution;

impl Solution {
    pub fn is_anagram(s: String, t: String) -> bool {
        if s.len() != t.len() {
            return false;
        }
        let mut s_chars = s.chars().collect::<Vec<char>>();
        let mut t_chars = t.chars().collect::<Vec<char>>();
        s_chars.sort();
        t_chars.sort();
        s_chars == t_chars
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_aa_a() {
        assert!(!Solution::is_anagram("aa".into(), "a".into()));
    }

    #[test]
    fn test_a_b() {
        assert!(!Solution::is_anagram("a".into(), "b".into()));
    }

    #[test]
    fn test_anagram_nagaram() {
        assert!(Solution::is_anagram("anagram".into(), "nagaram".into()));
    }

    #[test]
    fn test_rat_car() {
        assert!(!Solution::is_anagram("rat".into(), "car".into()));
    }
}
